applications/convection_diffusion_application/python_scripts/convection_diffusion_base_solver.py

```diff
@@ -0,0 +1,757 @@
+    _create_convergence_criterion
+    _create_linear_solver
+    _create_builder_and_solver
+    _create_convection_diffusion_solution_strategy
+    _create_restart_utility
+
+    The convection_diffusion_solution_strategy, builder_and_solver, etc. should alway be retrieved
+    using the getter functions get_convection_diffusion_solution_strategy, get_builder_and_solver,
+    etc. from this base class.
+
+    Only the member variables listed below should be accessed directly.
+
+    Public member variables:
+    model -- the model containing the modelpart used to construct the solver.
+    settings -- Kratos parameters containing solver settings.
+    """
+    def __init__(self, model, custom_settings):
+        super(ConvectionDiffusionBaseSolver, self).__init__(model, custom_settings)
+
+        default_settings = KratosMultiphysics.Parameters("""
+        {
+            "model_part_name" : "Thermic",
+            "domain_size" : -1,
+            "echo_level": 0,
+            "buffer_size": 2,
+            "analysis_type": "non_linear",
+            "model_import_settings": {
+                "input_type": "mdpa",
+                "input_filename": "unknown_name"
+            },
+            "restart_settings" : {
+                "load_restart"  : false,
+                "save_restart"  : false
+            },
+            "computing_model_part_name" : "ThermicModelPart",
+            "material_import_settings" :{
+                "materials_filename": ""
+            },
+            "convection_diffusion_variables" : {
+                "density_variable"              : "DENSITY",
+                "diffusion_variable"            : "CONDUCTIVITY",
+                "unknown_variable"              : "TEMPERATURE",
+                "volume_source_variable"        : "HEAT_FLUX",
+                "surface_source_variable"       : "FACE_HEAT_FLUX",
+                "projection_variable"           : "PROJECTED_SCALAR1",
+                "convection_variable"           : "CONVECTION_VELOCITY",
+                "mesh_velocity_variable"        : "MESH_VELOCITY",
+                "transfer_coefficient_variable" : "",
+                "velocity_variable"             : "VELOCITY",
+                "specific_heat_variable"        : "SPECIFIC_HEAT",
+                "reaction_variable"             : "REACTION_FLUX"
+            },
+            "time_stepping" : { },
+            "reform_dofs_at_each_step": false,
+            "line_search": false,
+            "compute_reactions": true,
+            "block_builder": true,
+            "clear_storage": false,
+            "move_mesh_flag": false,
+            "convergence_criterion": "residual_criterion",
+            "solution_relative_tolerance": 1.0e-4,
+            "solution_absolute_tolerance": 1.0e-9,
+            "residual_relative_tolerance": 1.0e-4,
+            "residual_absolute_tolerance": 1.0e-9,
+            "max_iteration": 10,
+            "linear_solver_settings":{
```

Comment:
i would use the amgcl solver as deafult solver. It will be much more robust than this choice